app/src/main/java/com/github/se/project/ui/components/LessonLocationDisplay.kt

```diff
@@ -23,33 +49,130 @@ fun LessonLocationDisplay(
+                          zoomControlsEnabled = false,
+                          scrollGesturesEnabled = false,
+                          zoomGesturesEnabled = false,
+                          tiltGesturesEnabled = false,
+                          rotationGesturesEnabled = false)) {
+                    // Marker for the lesson location
+                    Marker(state = MarkerState(position = lessonLocation), title = lessonTitle)
+                  }
+            }
+          }
+    }
+  }
+}
+
+/**
+ * Utility function to calculate the bounds to show both the user's location and lesson location,
+ * and update the camera position to fit them.
+ */
+fun getCameraPositionForBothLocations(
+    userLocation: LatLng?,
+    lessonLocation: LatLng,
+): CameraPosition {
+  // If we don't have the user's location, just focus on the lesson location
+  if (userLocation == null) {
+    return CameraPosition.fromLatLngZoom(lessonLocation, 10f)
+  }
+
+  // Otherwise, calculate the bounds to fit both locations
+  val latitudes = listOf(userLocation.latitude, lessonLocation.latitude)
+  val longitudes = listOf(userLocation.longitude, lessonLocation.longitude)
+
+  // Get the bounds to show both locations
+  val latMin = latitudes.minOrNull() ?: userLocation.latitude
+  val latMax = latitudes.maxOrNull() ?: userLocation.latitude
+  val lonMin = longitudes.minOrNull() ?: userLocation.longitude
+  val lonMax = longitudes.maxOrNull() ?: userLocation.longitude
+
+  // Create the bounds
+  val bounds = LatLngBounds(LatLng(latMin, lonMin), LatLng(latMax, lonMax))
+
+  // Calculate the zoom level based on the distance between the two locations
+  val distance = calculateDistance(userLocation, lessonLocation)
+  val zoomLevel = adjustZoomBasedOnDistance(distance)
+
+  // Adjust the camera to fit both locations with appropriate zoom
+  return CameraPosition.fromLatLngZoom(bounds.center, zoomLevel)
+}
+
+/** Private function to calculate the distance (in meters) between two locations. */
+fun calculateDistance(userLocation: LatLng?, lessonLocation: LatLng): Float {
+  if (userLocation == null) return 0f
+
+  val lat1 = userLocation.latitude
+  val lon1 = userLocation.longitude
+  val lat2 = lessonLocation.latitude
+  val lon2 = lessonLocation.longitude
+
+  val earthRadius = 6371 // Radius of the Earth in kilometers
+  val latDiff = Math.toRadians(lat2 - lat1)
+  val lonDiff = Math.toRadians(lon2 - lon1)
+
+  val a =
+      sin(latDiff / 2).pow(2.0) +
+          cos(Math.toRadians(lat1)) * cos(Math.toRadians(lat2)) * sin(lonDiff / 2).pow(2.0)
+  val c = 2 * atan2(sqrt(a), sqrt(1 - a))
+
+  // Distance in kilometers
+  val distanceInKm = earthRadius * c
+
+  // Convert to meters
+  return (distanceInKm * 1000).toFloat()
+}
+
+/**
+ * Private function to adjust the zoom level based on the distance between two locations. The
+ * further the locations, the lower the zoom level.
+ */
+fun adjustZoomBasedOnDistance(distance: Float): Float {
```

Comment:
As mentioned above, it can be a little too zoomed-out for some medium distances like what I tested (EPFL-Vidy). Maybe introducing other steps between 500 and 2000 meters, and between 2000 and 10000 could help?